trailofbits/deepstate
# Frontend fixing

Quite a lot of changes. But mostly fixes and enhancements, general idea and most of the code stay the same.

* Make frontends (executors) arguments uniform. Meaning that they share all the basic flags and when some specific fuzzer doesn't support some flag, the flag is just skipped. Only important exceptions are made if necessary, i.e. for angora second, taint binary.
* Add support for `--target_args` (flags passed to deepstate harness) and `--fuzzer_args` (flags passed to the backend fuzzer). Use them like `--target_args a=b abc=x` -> `-a b --abc x`
* Arguments provided by fuzz.py (including inherited from base.py):
  ```
  Guaranteed arguments (have default value):
    - output_test_dir (default: out)
    - mem_limit (default: 50MiB)
    - max_input_size (default: 8192B)
    - fuzzer_args (default: {})
    - blackbox (default: False)
    - post_stats (default: False)

  Optional arguments (may be None):
    - input_seeds
    - dictionary
    - exec_timeout

  Arguments that should not be used by child class:
    - timeout (default: 0)
    - num_workers (default: 1)
    - target_args (default: {})
  ```
* Add few `pre_exec` checks and add some more error handling
* Add `main` method to fuzz.py (since every backend fuzzer main method looks the same)

TODO:
* do something with fuzzer-compilations (json databases as suggested?)
* review `run` method
* pretty printing of statistic in single mode
